doc: add k8s vault-helm doc (#7193)
hashicorp/vault: website/source/docs/platform/k8s/helm.html.md

---
layout: "docs"
page_title: "Helm - Kubernetes"
sidebar_current: "docs-platform-k8s-helm"
description: |-
  The Vault Helm chart is the recommended way to install and configure Vault on Kubernetes.
---

# Helm Chart

The [Vault Helm chart](https://github.com/hashicorp/vault-helm)
is the recommended way to install and configure Vault on Kubernetes.
In addition to running Vault itself, the Helm chart is the primary
method for installing and configuring Vault to integrate with other
services such as Consul for High Availability deployments.

This page assumes general knowledge of [Helm](https://helm.sh/) and
how to use it. Using Helm to install Vault will require that Helm is
properly installed and configured with your Kubernetes cluster.

-> **Important:** The Helm chart is new and
may still change significantly over time. Please always run Helm with
`--dry-run` before any install or upgrade to verify changes.

> ~> **Security Warning:** TODO

Reviewer: Still `TODO`?

Author: Good catch, added one!

---
layout: "docs"
page_title: "Helm - Kubernetes"
sidebar_current: "docs-platform-k8s-helm"
description: |-
  The Vault Helm chart is the recommended way to install and configure Vault on Kubernetes.
---

# Helm Chart

The [Vault Helm chart](https://github.com/hashicorp/vault-helm)
is the recommended way to install and configure Vault on Kubernetes.
In addition to running Vault itself, the Helm chart is the primary
method for installing and configuring Vault to integrate with other
services such as Consul for High Availability deployments.

This page assumes general knowledge of [Helm](https://helm.sh/) and
how to use it. Using Helm to install Vault will require that Helm is
properly installed and configured with your Kubernetes cluster.

-> **Important:** The Helm chart is new and
may still change significantly over time. Please always run Helm with
`--dry-run` before any install or upgrade to verify changes.

~> **Security Warning:** By default, the chart will install an insecure configuration

> of Vault. This provides a less complicated out-of-box experience for new users,

Reviewer: Do we want to link to a "production ready" example here?

but is not appropriate for a production setup. It is highly recommended to use

> a properly secured Kubernetes cluster.

Reviewer: Do we want to link to something like https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/ here?

      - <a name="v-server-auditstorage-enabled" href="#v-server-auditstorage-enabled">`enabled`</a> (`boolean: true`) -
      Enables a persistent volume to be created for storing Vault's audit logs.

      - <a name="v-server-auditstorage-size" href="#v-server-auditstorage-size">`size`</a> (`string: 10Gi`) -
      Size of the volume to be created for Vault's audit logs.

      - <a name="v-server-auditstorage-storageclass" href="#v-server-auditstorage-storageclass">`storageClass`</a> (`string: null`) -
      Name of the storage class to use when creating the audit storage volume.

      - <a name="v-server-auditstorage-accessmode" href="#v-server-auditstorage-accessmode">`accessMode`</a> (`string: ReadWriteOnce`) -
      Type of access mode of the storage device.

  * <a name="v-server-dev" href="#v-server-dev">`dev`</a> - This configures `dev` mode for the Vault server.

      - <a name="v-server-dev-enabled" href="#v-server-dev-enabled">`enabled`</a> (`boolean: false`) -
      Enables `dev` mode for the Vault server.  This mode is useful for experimenting with Vault without needing to unseal.

>         ~> **Security Warning:** Never, ever, ever run a "dev" mode server in production. It is insecure and will lose data on every restart (since it stores data in-memory). It is only made for development or experimentation.

Reviewer: Maybe link to a "production ready" section with a reference architecture or check list for the making this production ready?

method for installing and configuring Vault to integrate with other
services such as Consul for High Availability deployments.

This page assumes general knowledge of [Helm](https://helm.sh/) and
how to use it. Using Helm to install Vault will require that Helm is
properly installed and configured with your Kubernetes cluster.

-> **Important:** The Helm chart is new and
may still change significantly over time. Please always run Helm with
`--dry-run` before any install or upgrade to verify changes.

~> **Security Warning:** By default, the chart will install an insecure configuration
of Vault. This provides a less complicated out-of-box experience for new users,
but is not appropriate for a production setup. It is highly recommended to use
a [properly secured Kubernetes cluster](https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/). 
See the [architecture reference](/docs/platform/k8s/run.html#architecture) 
for a Vault Helm production deployment checklist.

## Using the Helm Chart

To use the Helm chart, you must download or clone the
[vault-helm GitHub repository](https://github.com/hashicorp/vault-helm)

> and run Helm against the directory. We plan to transition to using a real

Reviewer 2: Maybe leave out the speculative "coming soon" plan.

Helm repository soon. When running Helm, we highly recommend you always
checkout a specific tagged release of the chart to avoid any
instabilities from master.

Prior to this, you must have Helm installed and configured both in your
Kubernetes cluster and locally on your machine. The steps to do this are

> out of the scope of this document, please read the

Reviewer 2: ```suggestion
out of the scope of this document. Please refer to the
```

[Helm documentation](https://helm.sh/) for more information.

Example chart usage:

```sh
# Clone the chart repo
$ git clone https://github.com/hashicorp/vault-helm.git
$ cd vault-helm

# Checkout a tagged version
$ git checkout v0.1.0

# Run Helm
$ helm install --dry-run ./
```

## Configuration (Values)

The chart is highly customizable using
[Helm configuration values](https://docs.helm.sh/using_helm/#customizing-the-chart-before-installing).

> Each value has a sane default tuned for an optimal getting started experience

Reviewer 2: ```suggestion
Each value has a default tuned for an optimal getting started experience
```

with Vault. Before going into production, please review the parameters below
and consider if they're appropriate for your deployment.

> * <a name="v-global" href="#v-global">`global`</a> - These global values affect multiple components of the chart.

Reviewer 2: Are these anchors required? I thought the site generator automatically created them? e.g. https://www.vaultproject.io/api/auth/jwt/index.html#oidc_discovery_url (the md has nothing special to achieve this).

You can test this by reviewing the preview build on netlify.

>   * <a name="v-global-enabled" href="#v-global-enabled">`enabled`</a> (`boolean: true`) - The master enabled/disabled configuration. If this is true, most components will be installed by default. If this is false, no components will be installed by default and manually opt-in is required, such as by setting <a href="#v-">`server.enabled`</a> to true.

Reviewer 2: ```suggestion
  * <a name="v-global-enabled" href="#v-global-enabled">`enabled`</a> (`boolean: true`) - The master enabled/disabled configuration. If this is true, most components will be installed by default. If this is false, no components will be installed by default and manually opting-in is required, such as by setting <a href="#v-">`server.enabled`</a> to true.
```

      If true, then the agent will be configured to automatically load HCL/JSON
      configuration files from this volume with `-config-dir`. This defaults
      to false.

        ```yaml
        extraVolumes:
          -  type: "secret"
             name: "consul-certs"
             load: false
        ```

  * <a name="v-server-affinity" href="#v-server-affinity">`affinity`</a> (`string`) - This value defines the [affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) for server pods. It defaults to allowing only a single pod on each node, which minimizes risk of the cluster becoming unusable if a node is lost. If you need to run more pods per node (for example, testing on Minikube), set this value to `null`.

    ```yaml
    # Recommended default server affinity:
    affinity: |
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
          matchLabels:
            app: {{ template "vault.name" . }}
            release: "{{ .Release.Name }}"
            component: server
          topologyKey: kubernetes.io/hostname
    ```

>   * <a name="v-server-service" href="#v-server-service">`extraVolumes`</a> - This configures the `Service` resource create for the Vault server.

Reviewer 2: ```suggestion
  * <a name="v-server-service" href="#v-server-service">`extraVolumes`</a> - This configures the `Service` resource created for the Vault server.
```

      - <a name="v-server-standalone-enabled" href="#v-server-standalone-enabled">`enabled`</a> (`boolean: true`) -
      Enables `standalone` mode for the Vault server.  This mode uses the `file` storage backend and requires a volume for persistence (`dataStorage`).

      - <a name="v-server-standalone-config" href="#v-server-standalone-config">`config`</a> (`string: "{}"`) -
      A raw string of extra HCL or JSON [configuration](https://www.vaultproject.io/docs/configuration/index.html) for Vault servers.
      This will be saved as-is into a ConfigMap that is read by the Vault servers.
      This can be used to add additional configuration that isn't directly exposed by the chart.

        ```yaml
        # ExtraConfig values are formatted as a multi-line string:
        config: |
          api_addr = "http://POD_IP:8200"

          listener "tcp" {
            tls_disable = 1
            address     = "0.0.0.0:8200"
          }

          storage "file" {
            path = "/vault/data"
          }
        ```

        This can also be set using Helm's `--set` flag (vault-helm v0.1.0 and later), using the following syntax:

        ```shell

>         --set 'server.standalone.config='{ listener "tcp" { address = "0.0.0.0:8200" }'

Reviewer 2: ```suggestion
        --set server.standalone.config='{ listener "tcp" { address = "0.0.0.0:8200" }'
```

      - <a name="v-server-ha-config" href="#v-server-ha-config">`config`</a> (`string: "{}"`) -
      A raw string of extra HCL or JSON [configuration](https://www.vaultproject.io/docs/configuration/index.html) for Vault servers.
      This will be saved as-is into a ConfigMap that is read by the Vault servers.
      This can be used to add additional configuration that isn't directly exposed by the chart.

        ```yaml
        # ExtraConfig values are formatted as a multi-line string:
        config: |
          ui = true
          api_addr = "http://POD_IP:8200"
          listener "tcp" {
              tls_disable = 1
              address     = "0.0.0.0:8200"
          }

          storage "consul" {
              path = "vault"
              address = "HOST_IP:8500"
          }
        ```

        This can also be set using Helm's `--set` flag (vault-helm v0.1.0 and later), using the following syntax:

        ```shell

>         --set 'server.ha.config='{ listener "tcp" { address = "0.0.0.0:8200" }'

Reviewer 2: ```suggestion
        --set server.ha.config='{ listener "tcp" { address = "0.0.0.0:8200" }'
```

* <a name="v-ui" href="#v-ui">`ui`</a> - Values that configure the Vault UI.

  - <a name="v-ui-enabled" href="#v-ui-enabled">`enabled`</a> (`boolean: false`) - If true, the UI will be enabled. The UI will only be enabled on Vault servers. If `server.enabled` is false, then this setting has no effect. To expose the UI in some way, you must configure `ui.service`.

  - <a name="v-ui-servicetype" href="#v-ui-servicetype">`serviceType`</a> (`string: ClusterIP`) -
  The service type to register. This defaults to `ClusterIP`.
  The available service types are documented on
  [the Kubernetes website](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types).

## Using the Helm Chart to Deploy Vault Enterprise

You can also use this Helm chart to deploy Vault Enterprise by following a few extra steps.

Find the license file that you received in your welcome email. It should have the extension `.hclic`. You will use the contents of this file to install the license in Vault.

In your `values.yaml`, change the value of `global.image` to one of the enterprise [release tags](https://hub.docker.com/r/hashicorp/vault-enterprise/tags).

```yaml
global:

>   image: "hashicorp/vault-enterprise:1.2.0-beta2"

Reviewer 2: ```suggestion
  image: "hashicorp/vault-enterprise:1.2.0"
```
Though 1.2.1 is probably even better...

Reviewer 2: Same change in other places after this.

[vault-helm GitHub repository](https://github.com/hashicorp/vault-helm)
and run Helm against the directory. We plan to transition to using a real
Helm repository soon. When running Helm, we highly recommend you always
checkout a specific tagged release of the chart to avoid any
instabilities from master.

Prior to this, you must have Helm installed and configured both in your
Kubernetes cluster and locally on your machine. The steps to do this are
out of the scope of this document. Please refer to the
[Helm documentation](https://helm.sh/) for more information.

Example chart usage:

```sh
# Clone the chart repo
$ git clone https://github.com/hashicorp/vault-helm.git
$ cd vault-helm

# Checkout a tagged version
$ git checkout v0.1.0

# Run Helm
$ helm install --dry-run ./
```

## Configuration (Values)

The chart is highly customizable using
[Helm configuration values](https://docs.helm.sh/using_helm/#customizing-the-chart-before-installing).
Each value has a default tuned for an optimal getting started experience
with Vault. Before going into production, please review the parameters below
and consider if they're appropriate for your deployment.

> * global` - These global values affect multiple components of the chart.

Reviewer 3: There appears to be a missing ` here.

Author: Good catch @jbayer, thanks!
